mlflow/pyfunc/__init__.py

```diff
@@ -1054,6 +1179,44 @@ def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LO
 
     model_metadata = Model.load(os.path.join(local_model_path, MLMODEL_FILE_NAME))
 
+    model_output_schema = model_metadata.get_output_schema()
+
+    if result_type is None:
+        if model_output_schema is None:
+            _logger.warning(
+                "No 'result_type' provided for spark_udf and the model metadata does not "
+                "have an output schema. 'result_type' is set to 'double' type."
+            )
+            result_type = "double"
+        else:
+            result_type = _infer_spark_udf_return_type(model_output_schema)
+
+    result_type = "boolean" if result_type == "bool" else result_type
+
+    if not isinstance(result_type, SparkDataType):
+        result_type = _parse_spark_datatype(result_type)
+
+    if not _check_udf_return_type(result_type):
+        raise MlflowException(
+            f"""Invalid 'spark_udf' result type: {result_type}.
+'spark_udf' return type must be one of the following types:
+Primitive types:
+ - int
+ - long
+ - float
+ - double
+ - string
+ - boolean
+or compound types:
+ - array<primitive>: An array of primitives, e.g., array<int>.
+ - array<array<primitive>>: A double nested array type consists of primitive type elements,
+   e.g., array<array<int>>.
+ - struct<...>: A struct type, its field type can be primitive, array<primitive>,
+   or array<array<primitive>>, e.g., struct<a:int,b:array<int>>.
+ - array<struct<...>>: An array of structs, e.g., array<struct<a:int,b:array<int>>>.
```

Comment:
```suggestion
 - array<primitive>: ...
 - array<array<primitive>>: ...
 - array<struct<primitive>>
 - struct<primitive>: ...
 - struct<array<primitive>>: ...
 - struct<array<array<primitive>>>: ...
```


This is what I meant by listing all possible types.

Reply:
Not sure if this covers everything.

Reply:
`struct<primitive>` is not a valid type string, we should use format like
`struct<field1: type1, field2: type2, ...>`

Reply:
Got it. Can we ues `struct<field: primitive | array<primitive> | array<array<primitive>>, ...>`?